pkg/process/metadata/parser/java/util.go

```diff
@@ -0,0 +1,126 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package javaparser
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/rickar/props"
+	"github.com/vibrantbyte/go-antpath/antpath"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type mapSource struct {
+	m map[string]string
+}
+
+// Get the value for the supplied key
+func (y *mapSource) Get(key string) (string, bool) {
+	val, ok := y.m[key]
+	return val, ok
+}
+
+// GetDefault gets the value for the supplied key or the defVal if missing
+func (y *mapSource) GetDefault(key string, defVal string) string {
+	val, ok := y.m[key]
+	if !ok {
+		return defVal
+	}
+	return val
+}
+
+func newArgumentSource(arguments []string, prefix string) *mapSource {
+	parsed := make(map[string]string)
+	for _, val := range arguments {
+		if strings.HasPrefix(val, prefix) {
+			parts := strings.SplitN(val[len(prefix):], "=", 2)
+			if len(parts) == 1 {
+				parsed[parts[0]] = ""
+			} else {
+				parsed[parts[0]] = parts[1]
+			}
+			continue
+		}
+	}
+	return &mapSource{parsed}
+}
+
+func newPropertySourceFromStream(rc io.ReadCloser, filename string) (*props.PropertyGetter, error) {
+	var properties props.PropertyGetter
+	var err error
+	if strings.HasSuffix(filename, ".properties") {
+		properties, err = props.Read(rc)
+	} else if strings.HasSuffix(filename, ".yaml") || strings.HasSuffix(filename, ".yml") {
+		properties, err = newYamlSource(rc)
+	} else {
+		err = fmt.Errorf("unhandled file type for %s", filename)
+	}
+	return &properties, err
+}
+
+func longestPathPrefix(pattern string) string {
+	idx := strings.IndexAny(pattern, "?*")
+	if idx < 0 {
+		return pattern
+	}
+	idx = strings.LastIndex(pattern[:idx], "/")
+	if idx < 0 {
+		return ""
+	}
+	return pattern[:idx]
+}
```

Comment:
not sure what you're doing here or what's the purpose

Reply:
I'm extracting the longest path that's not a pattern in order to start waking the file system  from the best point I can to avoid doing ant pattern matching too many times

For instance if I've `/opt/somewhere/**/application.properties` I will start waking from `/opt/somewhere` and not from the root saving few pattern maching to enter into /opt and then /opt/somewhere